weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/WeldDeploymentCleanupProcessor.java

```diff
@@ -74,39 +71,26 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
             return;
         }
 
-        // extract all service controllers and store them as attachment for WeldDeploymentCleanupProcessor to use
-        List<ServiceController> serviceControllers = new ArrayList<>();
-
-        // add controllers from top level deployment
-        for (ServiceName serviceName : getComponentStartServiceNames(deploymentUnit)) {
-            ServiceController<?> controller = deploymentUnit.getServiceRegistry().getService(serviceName);
-            if (controller != null) {
-                serviceControllers.add(controller);
-            }
+        // add dependency on our WeldBootstrapService and WeldStartService to ensure this goes after them
+        ServiceBuilder<?> weldStartCompletionServiceBuilder = serviceTarget.addService(weldStartCompletionServiceName);
+        final Supplier<WeldBootstrapService> bootstrapSupplier = weldStartCompletionServiceBuilder.requires(weldBootstrapServiceName);
+        weldStartCompletionServiceBuilder.requires(weldStartServiceName);
+        // require component start services from top level deployment
+        for (ServiceName componentStartSN : getComponentStartServiceNames(deploymentUnit)) {
+            weldStartCompletionServiceBuilder.requires(componentStartSN);
```

Comment:
Why didn't this use service dependencies before instead of this StabilityMonitor thing? It seems like the obvious and far easier thing to have done.

There was something complicated about this service but I don't recall what it was.

Reply:
@ropalka ^^^

Reply:
https://issues.redhat.com/browse/WFLY-10784 is why. I don't completely understand but the switch from dependencies was deliberate.

Reply:
Anything depending on StabilityMonitors is broken @bstansberry @manovotn 
because StabilityMonitors might report false positives about achieved stability
(see https://issues.redhat.com/browse/MSC-255 ).
AFAICS WeldDeploymentCleanupProcessor is depending just on EE_MODULE_DESCRIPTOR
start component controllers and these should be always available (are not optional) if component is registered.
Further more I see there was a test case provided for WFLY-10784 and it is passing.

Reply:
Hmm, the switch was made because we needed to make sure our service goes last, even after optional deployments (which was the cause of the problem in the linked JIRA). I think the problem was that when the optional service failed to boot, Weld was stuck waiting on it. The stability monitor was a suggested way to fix that issue which seemed to have worked.
That being said, I know little of MSC, so I just went with the suggestion back then. There could have been a different way to achieve the same thing.

Reply:
If setting the dependencies does the trick of Weld going last, then it's probably fine. I am no MSC expert nor do I have a memory of why are all those services exactly this way. So let's just say that green tests are a good sign and in favor of accepting the change ;-)

Reply:
Thanks @manovotn. It's true the test is passing, so this seems good. 👍